### DISCUSSION TITLE
Lesson 12 Invariants Test:  Redeem collateral failing with error:  Arithmetic over/underflow

### DISCUSSION THREAD

**Discussion by jojomensah89** [Cyfrin/foundry-full-course-cu#601] [Q&A] [answered]
I get the error  Arithmetic over/underflow when I run the invariants tests.
This is the code for the test in Handler.t.sol
```
 function redeemCollateral(uint256 _collateralSeed, uint256 _amountCollateral) public {
        ERC20Mock collateral = _getCollateralFromSeed(_collateralSeed);
        uint256 maxCollateralToRedeem = zaxEngine.getCollateralBalanceOfUser(msg.sender, address(collateral));
        _amountCollateral = bound(_amountCollateral, 0, maxCollateralToRedeem);

        if (_amountCollateral == 0) {
            return;
        }
        zaxEngine.redeemCollateral(address(collateral), _amountCollateral);
    }
```
And when I run 
```
forge test --mt invariant_protocolMustHaveMoreValueThanTotalSupply -vvvv
```
I get this error
```
[FAIL. Reason: Arithmetic over/underflow]
        [Sequence]
                sender=0x000000000000000000000000000000000000000a addr=[test/fuzz/Handler.t.sol:Handler]0x2e234dae75c793f67a35089c9d99245e1c58470b calldata=redeemCollateral(uint256,uint256), args=[223610849192816100937918987304 [2.236e29], 454035980949209708232323300959 [4.54e29]]
                sender=0x0000000000000000000000000000000000000b7f addr=[test/fuzz/Handler.t.sol:Handler]0x2e234dae75c793f67a35089c9d99245e1c58470b calldata=failed():(bool), args=[]
```
Any help will be much appreciated

**Comment by theirrationalone** [comment, marked as answer]
We need a random sender to sign & send the tx... 


try this:

```solidity
vm.prank(msg.sender) ;
zaxEngine.redeemCollateral(address(collateral);
```

**Reply by jojomensah89** [reply]
Yes😅
Thanks, It worked

**Reply by paprikrumplikas** [reply]
Hej @theirrationalone,

Could you please let me know why exactly we need that line? 
What address is calling the redeemCollateral function if I do not add that line? 
It is msg.sender either way, is not it?

**Reply by theirrationalone** [reply]
We say to foundry... Heyy foundry it's just a prank (running Or testing smc locally) please sponsor it... 

Foundry..... Okay but i wanna know who will exec this tx Or function???? If you don't specify the sender then i'll let the wrapping SMC be the sender of this tx Or please specify the sender of this prank tx by putting sender's address inside prank function. 


And @orgovaan ! Do you know who's the wrapping SMC??? 

It's the testing function, which hasn't deposited any amount.... Thus far! 


We could implement a bound & a check and could record senders who have deposited, to allow only those who have deposited some amount. 

In the above case of msg.sender has deposited so only that exact same sender could redeem and nobody else. That was the reason for the underflow (actually) exception. 

Redeem 0  - (amount > 0) .... 


Hope it helps to vanish the misconception.... Fuzz testing is awesome... 

First decide and guess the unbreakable condition and then ignite & flood to reveal the edge cases.... 


In fuzz testing we would need our expectations..... We should first learn how to perform a fuzz test and what're its caveats.... 

We should do only one thing PRACTICE PRACTICE PRACTICE.... and Should ask Qs here in the community if we feel exhausted. 

@orgovaan ! If still it's not cleared then please tell me i would help further... 

Sorry if i'm wrong here 😅

**Reply by paprikrumplikas** [reply]
Hej @theirrationalone,

Thanks for your answer. I still have some doubts.
I understand that by default, the transactions are sent by the testing smart contract... so the redeemCollateral() function is being called by the testing smart contract, meaning that msg.sender == the testing smart contract.
So in my mind adding vm.prank(msg.sender) should not change anything.
What am I missing here?

Cheers.

**Reply by theirrationalone** [reply]
I think handler.sol isn't the msg.sender in fuzz testing, msg. sender might be the invariant.sol instead..... 

I haven't experimented contextually yet so, it could be an opportunity for you to prove me wrong... 😅😅😅

**Reply by theirrationalone** [reply]
Or could be a random sender.... 😅

**Reply by paprikrumplikas** [reply]
Still... whatever address calls the redeemCollateral() function is the msg.sender. 
Thus, pranking msg.sender does not make any sense to me.

**Reply by theirrationalone** [reply]
It's arguably mine misconception..... Maybe 😅

But if you're unable to digest this trick then you can use a bound of addresses within that a recorder.... 

Recorder could be a mapping Or an array of addresses that would help you to not get into that underflow exception..... 

Record those addresses that have deposited... You can record them underneath the deposit function into handler.sol and then can use a bound to call redeemcollateral function conditionally....!

**Reply by eugenebuhor** [reply]
I also had the arithmetic over/underflow error thrown in `redeemCollateral` function when was running the invariant tests. And I did a research and found some details about it - it's potentially a misusing of `msg.sender` directly in the fuzz tests. Here's the link to the [discussion](https://github.com/Cyfrin/foundry-full-course-f23/discussions/1333) I created separately (it looks like an issue that needs to be discussed). Hopefully it helps.
